Fix window position on macOS (2.1)

rraallvv opened the pull request:
On macOS the window position is wrong when the game is launched from the editor, this is because the Y coordinate starts from the bottom on macOS. The problem is because Godot computes the `--position` parameter relative to the top-left corner with the Y coordinate starting from the top and increasing its value downwards. The modifications fix the problem for 2.1

Related to issue #7477

rraallvv committed:
fix window position computations on macOS

bojidar-bg commented:
Wouldn't this mean that `set_window_position(get_window_position())` would change the position of the window on OSX?

I think we should fix it in either `set_window_position` or the parsing of `--position`, not when we start the game..

rraallvv commented:
@bojidar-bg 

I think `set_window_position(get_window_position())` does "place" the window in the same position, i.e. the window is not moved.

These are both functions: 

	Point2 OS_OSX::get_window_position() const {
		NSRect nsrect = [window_object frame];

		Point2 pos;

		// Return the position of the top-left corner, for OS X the y starts at the bottom
		pos.x = nsrect.origin.x * display_scale;
		pos.y = (nsrect.origin.y + nsrect.size.height) * display_scale;

		return pos;
	};

	void OS_OSX::set_window_position(const Point2 &p_position) {

		NSPoint pos;

		pos.x = p_position.x / display_scale;
		pos.y = p_position.y / display_scale;

		[window_object setFrameTopLeftPoint:pos];

		_update_window();
	};


Performing both calls would be equivalent to this:

	// get_window_position
	NSRect nsrect = [window_object frame];

	Point2 p_position;

	// Return the position of the top-left corner, for OS X the y starts at the bottom
	p_position.x = nsrect.origin.x * display_scale;
	p_position.y = (nsrect.origin.y + nsrect.size.height) * display_scale;

	// set_window_position
	NSPoint pos;

	pos.x = p_position.x / display_scale;
	pos.y = p_position.y / display_scale;

	[window_object setFrameTopLeftPoint:pos];

After factoring out some vars one gets:

	// get_window_position
	NSRect nsrect = [window_object frame];

	// set_window_position
	NSPoint pos;

	pos.x = (nsrect.origin.x * display_scale) / display_scale;
	pos.y = ((nsrect.origin.y + nsrect.size.height) * display_scale) / display_scale;

	[window_object setFrameTopLeftPoint:pos];

That would be equivalent to, retrieving the window frame, computing the top left corner, and setting the value with `-setFrameTopLeftPoint:`

	// get_window_position
	NSRect nsrect = [window_object frame];

	// set_window_position
	NSPoint pos;

	pos.x = nsrect.origin.x;
	pos.y = nsrect.origin.y + nsrect.size.height;

	[window_object setFrameTopLeftPoint:pos];

As whether the fix should be in either `set_window_position` or the parsing of `--position`, I think the patch does both things, although I'm not sure what you mean by "when we start the game".

marcelofg55 commented:
Isn't the position already working fine with the actual code at OS.set_window_position? I remember I worked on this issue and fixed it directly on OS.set_window_position.

rraallvv commented:
@marcelofg55 I was having a similar problem as the OP in issue #7477, this is the arrangement I currently have:

![screen shot 2017-11-03 at 10 16 25 am](https://user-images.githubusercontent.com/1840524/32379595-528251c4-c084-11e7-96eb-eb6da588d831.png)

![screen shot 2017-11-03 at 10 17 20 am](https://user-images.githubusercontent.com/1840524/32379605-5972ba46-c084-11e7-9219-8393efbb03a9.png)


The following are `set_window_position()` and `get_window_position()` in the current implementation:

	Point2 OS_OSX::get_window_position() const {

		Size2 wp([window_object frame].origin.x, [window_object frame].origin.y);
		wp *= display_scale;
		return wp;
	};

	void OS_OSX::set_window_position(const Point2 &p_position) {

		Size2 scr = get_screen_size();
		NSPoint pos;

		pos.x = p_position.x / display_scale;
		// For OS X the y starts at the bottom
		pos.y = (scr.height - p_position.y) / display_scale;

		[window_object setFrameTopLeftPoint:pos];

		_update_window();
	};


Calling `set_window_position(get_window_position())` would be equivalent to this:

	// get_window_position

	Size2 p_position([window_object frame].origin.x, [window_object frame].origin.y);
	p_position *= display_scale;

	// set_window_position

	Size2 scr = get_screen_size();
	NSPoint pos;

	pos.x = p_position.x / display_scale;
	// For OS X the y starts at the bottom
	pos.y = (scr.height - p_position.y) / display_scale;

	[window_object setFrameTopLeftPoint:pos];


Substituting some values:


	Size2 scr = get_screen_size();
	NSPoint pos;

	pos.x = [window_object frame].origin.x * display_scale / display_scale;
	// For OS X the y starts at the bottom
	pos.y = (scr.height - [window_object frame].origin.y * display_scale) / display_scale;

	[window_object setFrameTopLeftPoint:pos];


That would leave us with this:


	Size2 scr = get_screen_size();
	NSPoint pos;

	pos.x = [window_object frame].origin.x;
	// For OS X the y starts at the bottom
	pos.y = scr.height / display_scale - [window_object frame].origin.y;

	[window_object setFrameTopLeftPoint:pos];


With that snipped of code the window will be moved to a different position that the current one. Also the parameter read from `game_window_placement/rect_custom_position` and the calculation of the margin when trying to center the window on the screen, both are positive values for X and Y, so adding those values to the screen position won't do, that's where the conditional in [`editor_run.cpp`](https://github.com/godotengine/godot/pull/12607/files#diff-e83058b522d9e9a961d66092bab9113fR106) comes in, so that the Y coordinate is computed subtracting the value from the screen Y coordinate.

marcelofg55 commented:
But if we fix this on `editor_run.cpp`, like this PR, the problem still remains if you use `OS.set_window_position` from a gdscript. I agree with @bojidar-bg that this should be fixed within `set_window_position`.

rraallvv commented:
@marcelofg55 oh, I see... `OS.set_window_position(OS.get_window_position())` doesn't move the window but `OS.set_window_position(some_point)` needs `some_point ` to be an absolute value.

So for instance `OS.set_window_position(OS.get_screen_position(OS.get_current_screen()))` would move the window so that its top-left corner would sit on the top-left corner for the current screen, while `OS.set_window_position(Vector2(0, 0))` would place the top-left corner of the window on the bottom-left corner for the screen with index 0 (on macOS).

If `set_window_position` and `get_window_position` need the parameter to be relative to the current screen, then the parameter `--position` has to be absolute, currently it's computed relative to the current screen. Trying to make both relative would need an additional parameter to pass the screen from the editor.

I guess that's what @bojidar-bg meant in the post above.

rraallvv commented:
@bojidar-bg and @marcelofg55, I reverted the changes in `editor_run.cpp` and added the stuff to fix the window positioning issues inside `os_osx.mm` (technically `os_osx.h` too)

First I needed to compute the top-left coordinate for the smaller rectangle that encompasses all the screens, then I had to use that as the relative origin from where all the screen positions and window positions are calculated.

Then I used the old position getter and setter as helper functions to translate between the native coordinates and the coordinates expected by Godot.

The new getter and setter use the top-left origin of the encompassing rectangle and the helper functions so that the y-coordinates start at the top and increase downwards.

Also added a callback that allows to compute the top-left origin only when there is a change in the display arrangement.

akien-mga commented:
There are too many commits and unrelated changes in this PR, please check if you can rebase it to a single commit following http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

rraallvv commented:
@akien-mga done!

rraallvv commented:
@akien-mga fixed typos and made corrections.
